python/pyarrow/_cuda.pyx

```diff
@@ -0,0 +1,760 @@
+
+import six
+from pyarrow.compat import tobytes
+from pyarrow.lib cimport *
+from pyarrow.includes.libarrow_cuda cimport *
+from pyarrow.lib import py_buffer, allocate_buffer
+cimport cpython as cp
+
+
+cdef class Context:
+    """ CUDA driver context.
+    """
+
+    def __cinit__(self, int device_number=0):
+        """Construct the shared CUDA driver context for a particular device.
+
+        Parameters
+        ----------
+        device_number : int
+          Specify the gpu device for which the CUDA driver context is
+          requested.
+
+        """
+        cdef CCudaDeviceManager* manager
+        check_status(CCudaDeviceManager.GetInstance(&manager))
+        cdef int n = manager.num_devices()
+        if device_number >= n or device_number < 0:
+            self.context.reset()
+            raise ValueError('device_number argument must be '
+                             'non-negative less than %s' % (n))
+        check_status(manager.GetContext(device_number, &self.context))
+        self.device_number = device_number
+
+    @staticmethod
+    def get_num_devices():
+        """ Return the number of GPU devices.
+        """
+        cdef CCudaDeviceManager* manager
+        check_status(CCudaDeviceManager.GetInstance(&manager))
+        return manager.num_devices()
+
+    @property
+    def device_number(self):
+        """ Return context device number.
+        """
+        return self.device_number
+
+    cdef void init(self, const shared_ptr[CCudaContext] &ctx):
+        self.context = ctx
+
+    @property
+    def bytes_allocated(self):
+        """ Return the number of allocated bytes.
+        """
+        return self.context.get().bytes_allocated()
+
+    def allocate(self, nbytes):
+        """ Allocate CUDA memory on GPU device for this context.
+
+        Parameters
+        ----------
+        nbytes : int
+          Specify the number of bytes to be allocated.
+
+        Returns
+        -------
+        buf : CudaBuffer
+          Allocated buffer.
+        """
+        cdef shared_ptr[CCudaBuffer] cudabuf
+        check_status(self.context.get().Allocate(nbytes, &cudabuf))
+        return pyarrow_wrap_cudabuffer(cudabuf)
+
+    def open_ipc_buffer(self, ipc_handle):
+        """ Open existing CUDA IPC memory handle
+
+        Parameters
+        ----------
+        ipc_handle : IpcMemHandle
+          Specify opaque pointer to CUipcMemHandle (driver API).
+
+        Returns
+        -------
+        buf : CudaBuffer
+          referencing device buffer
+        """
+        handle = pyarrow_unwrap_cudaipcmemhandle(ipc_handle)
+        cdef shared_ptr[CCudaBuffer] cudabuf
+        check_status(self.context.get().OpenIpcBuffer(handle.get()[0],
+                                                      &cudabuf))
+        return pyarrow_wrap_cudabuffer(cudabuf)
+
+    def device_buffer(self, object obj=None,
+                      int64_t offset=0, int64_t size=-1):
+        """Create device buffer from object data.
```

Comment:
In special case, `device_buffer(obj=None, size=<size>)` is exactly `allocate(<size>)`.
In all other cases, when `obj is not None`, `device_buffer` does allocation and initialization from offset, or returns a zero-copy slice (when obj is CudaBuffer).
So, `allocate` would not be appropriate. `device_buffer` would be more similar to `py_buffer`. 

Creating a `CudaBuffer` requires a context. For a better API, I propose introducing `context` keyword to `py_buffer` so that users could use `pa.py_buffer(<obj>, context=pa.cuda.Context())`.

Reply:
Yes... so what's the point of allowing `device_buffer(obj=None, size=...)` in addition to `allocate(size)`?

APIs with multiple ways of doing the same thing impose a cognitive burden. When I see this, I don't know if there's any difference between the two APIs, which one is recommended and why. Further more, one is a single verb ("allocate"), another a bunch of names ("device_buffer"). I would like to see this simplified and made more consistent. For example "new_buffer" for the uninitialized case and "buffer_from_data" for the case where we copy something.

Finally, I don't think something that generally returns a _new_ buffer should sometimes return a logical slice. It's a recipe for confusion.

> For a better API, I propose introducing context keyword to py_buffer so that users could use `pa.py_buffer(<obj>, context=pa.cuda.Context())`.

That's a bad idea. `py_buffer` creates a logical buffer, it doesn't copy data around.

Reply:
OK, `new_buffer` and `buffer_from_data` arguments make sense. In total, there are four methods of creating CudaBuffer instances:
1. `Context.new_buffer` - that's `allocate`
2. `Contex.buffer_from_data` - data is always copied (from host to device or from device to device)
3. `CudaBuffer.slice` - zero-copy slice
4. `Context.open_ipc_buffer` - create zero-copy buffer from ipc handle or from a buffer containing the ipc handle. Would it make sense to call it `buffer_from_ipc_handle` or similar?

About the suggestion: replace `py_buffer` with `as_buffer`, if that makes a difference, otherwise ignore.
Btw, `py_buffer(<Python bytes object>)` makes a copy, so it is not a pure logical buffer creator.

Reply:
`open_ipc_buffer` sounds fine to me.

> Btw, py_buffer(<Python bytes object>) makes a copy, so it is not a pure logical buffer creator.

I don't think so:
```python
>>> b = b"12345" * 50
>>> buf = pa.py_buffer(b)
>>> buf.address - id(b)
32
>>> id(b) < buf.address < id(b) + sys.getsizeof(b)
True
```
The Arrow buffer does point inside the bytes object (just at the start of its payload, which is allocated contiguously to the 4-words header).

Reply:
CC: @wesm 
This is very interesting. I thought Python bytes are immutable (hence the copy assumption) but pyarrow plays with fire here, IMHO:
```
>>> import pyarrow as pa
>>> import numpy as np
>>> a = b'123456'
>>> a[0] = 77 # bytes are immutable
...
TypeError: 'bytes' object does not support item assignment
>>> b = pa.py_buffer(a)  # but with pyarrow bytes can be changed in-situ
>>> arr = np.frombuffer(b, dtype=np.uint8)
>>> arr
array([49, 50, 51, 52, 53, 54], dtype=uint8)
>>> arr[0] = 66 # change immutable 'a' in place
>>> a
b'B23456'
>>> hash(a)
-4581532003987476523
>>> arr[0] = 77
>>> a
b'M23456'
>>> hash(a) # hash value is constant
-4581532003987476523
```
But note that numpy respects immutability:
```
>>> arr2 = np.frombuffer(a, dtype=np.uint8)
>>> arr2
array([77, 50, 51, 52, 53, 54], dtype=uint8)
>>> arr2[0] = 88
...
ValueError: assignment destination is read-only
```

Reply:
@pearu Can you open an issue for this?

Reply:
Sure: [ARROW-3228](https://issues.apache.org/jira/browse/ARROW-3228) .

Reply:
Renamed `device_buffer` to `buffer_from_data` and implemented copy-only functionality.
Renamed `allocate` to `new_buffer`.
Renamed `allocate_host_buffer` to `new_host_buffer`.